### PR TITLE
Task runner (mid-level API), part 1

### PR DESCRIPTION
**Pull request by soywod**
Split version of #165, part 1.

- Added core changes
- Added the new `Resolver` flow that aims to enqueue task AND get resolution straight. `Resolver::run_task` returns a new `Flow` called `TaskResolver` that can feed `Stream::progress`.
- Made sure `Scheduler` and `Resolver` are `Send`.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -81,12 +82,27 @@ impl ClientFlow {
...
             receive_state,
         }
     }
 
+    /// Changes the actual receive state codec with the
+    /// [`ResponseCodec`]'s one.
+    ///
+    /// The main use case of this function is straight after receiving
+    /// greeting, in order to replace the actual greeting codec by the
+    /// default response codec.
```
Comment (jakoschiko) [inline src/client.rs:97]: You don't have to do this. `ClientFlow` already does this by default. See https://github.com/duesee/imap-flow/blob/e24687913ba0761235e6f1e863b17c732ff8b807/src/client.rs#L153-L157

File: tasks/src/lib.rs [hunk 2]
```diff
@@ -26,11 +27,11 @@ use thiserror::Error;
 /// and move out uninteresting responses (returning `Some(...)`).
 ///
 /// If no active task is interested in a given response, we call this response "unsolicited".
-pub trait Task: 'static {
+pub trait Task: Send + 'static {
```
Comment (jakoschiko) [inline tasks/src/lib.rs:30]: Why is `Send` necessary?

**Comment by jakoschiko** [comment]
I don't understand why we need a `Resolver` that wraps the `Scheduler`.

Suggestion: Let's add this method to `Scheduler`:

```rust
impl Scheduler {
    pub fn run_task<'a, T: Task>(&'a mut self, task: T) -> TaskRunner<'a, T> {
        let task_handle = self.enqueue_task(task);
        TaskRunner { scheduler: self, task_handle }
    }
}
```

`TaskRunner` looks like this:

```rust
pub struct TaskRunner<'a, T: Task> {
    scheduler: &'a mut Scheduler,
    task_handle: TaskHandle<T>,
}

impl<'a, T> Flow for TaskRunner<'a, T> {
    type Event = T::Output;
    type Error = SchedulerError;

    fn enqueue_input(&mut self, bytes: &[u8]) {
        self.scheduler.enqueue_input(bytes);
    }

    fn progress(&mut self) -> Result<Self::Event, FlowInterrupt<Self::Error>> {
        todo!()
    }
}
```

Then we can use it like this:

```rust
async fn main() {
    let stream = TcpStream::connect("127.0.0.1:12345").await.unwrap();
    let mut stream = Stream::insecure(stream);

    let client = ClientFlow::new(ClientFlowOptions::default());
    let mut scheduler = Scheduler::new(client);

    let capabilities = stream
        .progress(scheduler.run_task(CapabilityTask::default()))
        .await
        .unwrap();

    println!("capabilities: {capabilities:?}");

    stream
        .progress(scheduler.run_task(AuthenticateTask::plain("alice", "pa²²w0rd")))
        .await
        .unwrap();

    stream
        .progress(scheduler.run_task(LogoutTask::default()))
        .await
        .unwrap();
}
```

If we want to run tasks in parallel, we could even provide something like this:

```rust
let (foo, bar) = stream
    .progress(scheduler.run_task_2(FooTask::default(), BarTask::default()))
    .await
    .unwrap();
```

And we could play a little bit with the API to make it more readable:

```rust
let foo = stream
    .progress(FooTask::default().run(scheduler)) // Swap order to reduce nesting
    .await
    .unwrap();
```

Pros:

- No additional `Resolver` type
- Instead of duplicating the API and adding many functions to `Resolver`, we can reuse the existing `Task` constructors

What do you think?

File: src/client.rs (hunk 1 above)
Comment (soywod) [reply src/client.rs:97]: Oh, great! I missed it, thank you.

File: tasks/src/lib.rs (hunk 2 above)
Comment (soywod) [reply tasks/src/lib.rs:30]: I test my changes directly from `email-lib`, and somehow it was complaining about `Send` missing. Since my use case is really basic (I believe it should be the case of most users), I thought it would not hurt to add the `Send` restriction.

**Comment by soywod** [comment]
I initially tried sth similar, but the compiler was not happy at all about the lifetime introduction. Let me try again.

**Comment by soywod** [comment]
Awesome, looks like it work well this time. I did not think about creating a new struct *from the scheduler itself* that returns self, somehow the compiler is happy. Thank you, I prefer this version more!

**Comment by soywod** [comment]
(Regarding the CI, I wait for #175 to be merged to fix it)

File: tasks/src/lib.rs (hunk 2 above)
Comment (jakoschiko) [reply tasks/src/lib.rs:32]: > I test my changes directly from `email-lib`, and somehow it was complaining about `Send` missing.

I think you can write `T: Task + Send` in your lib in places where it's necessary.

> Since my use case is really basic (I believe it would be the case of most users), I thought it would not hurt to add the `Send` restriction since `this trait is automatically implemented when the compiler determines it’s appropriate` [[sources](https://doc.rust-lang.org/std/marker/trait.Send.html)].

Yes, all current implementors of `Task` are `Send`. But adding this bound

- implies that `imap-tasks` somehow relies on this bound, which is not true and
- prevents us from adding implementors in the future that are not `Send` (e.g. because they borrow data instead of owning them).

If there is still a good reason to add this bound, than I'm okay with it. But if you can solve your problem by writing `T: Task + Send` in `email-lib`, then I think we should prefer that. If you point me to your problem in `email-lib` then maybe I can better understand the nature of the problem.

**Comment by jakoschiko** [comment]
I'm on vacation for a couple of days. I'll provide some feedback next week.

File: tasks/src/lib.rs (hunk 2 above)
Comment (soywod) [reply tasks/src/lib.rs]: I dug a bit more, and the issue comes from places where `Box<dyn Any>` is used, mostly `TaskAny::process_tagged` and `TaskMap::tasks`. In `email-code`, the backend struct that holds a scheduler (soon a client) have the `Send` restriction by nature, which "imposes" this restriction down. From `email-lib` I have no control over this behaviour. Here the minimum changes I needed to do in order to make the compiler happy: https://github.com/duesee/imap-flow/pull/174/commits/0da136d34d08adc455dbae472eecc6d49b796c69.

I don't see any happy path:

- Without `Send` restriction in `imap-tasks`, users working with `Send` restriction cannot use the Scheduler
- With `Send` restriction in `imap-tasks`, users working without `Send` restriction (`Rc`) cannot use the Scheduler
- Users cannot change this behaviour from their side

Any idea @duesee?

Comment (jakoschiko) [reply tasks/src/lib.rs]: So your problem is that `Scheduler` is not `Send`. I think that's a good reason to add the bound to `Task`. Then we should also add an [assertion](https://docs.rs/static_assertions/latest/static_assertions/macro.assert_impl_all.html) to ensure that `Scheduler` is `Send` and we don't break it accidentally.

File: tasks/src/lib.rs [hunk 3]
```diff
@@ -342,6 +380,52 @@ pub enum SchedulerError {
...
+    UnexpectedByeResponse(Bye<'static>),
+
+    #[error("unexpected BAD response: {}", .0.text)]
+    UnexpectedBadResponse(StatusBody<'static>),
+
+    #[error("unexpected NO response: {}", .0.text)]
+    UnexpectedNoResponse(StatusBody<'static>),
+
+    #[error("missing required data for command {0}")]
+    MissingData(String),
```
Comment (duesee) [inline tasks/src/lib.rs:394]: I think we need to be a bit careful here to avoid already getting in the weeds of IMAP command semantics.

`UnexpectedTaggedResponse` is a `SchedulerError` because it signals that `imap-tasks` lost track of connection understanding. This could be a severe error during scheduling.

The variants we want to add here are rather task specific, though. Maybe we should create a dedicated enum that holds these common errors? This would also allow us to do refinements (as explained above).

I'm not sure yet about `UnexpectedByeResponse`...

File: tasks/src/tasks.rs
Comment (duesee) [inline tasks/src/tasks.rs:1]: Hehe, we should have started with dedicated files per task :-) Thanks for sorting this out!

What I intend to do: I want to move all tasks into dedicated files first, and then apply your changes. This way it's easier to see what has changed exactly. Unfortunately, we have no diffing tool yet that is smart enough to detect moves :-( Makes it rather hard to see what changed.

File: tasks/src/tasks/authenticate.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                },
+            )),
+            StatusKind::No => Err(SchedulerError::UnexpectedNoResponse(status_body)),
+            StatusKind::Bad => Err(SchedulerError::UnexpectedBadResponse(status_body)),
+        }
+    }
+}
+
+impl AuthenticateTask {
+    pub fn new_plain(login: &str, passwd: &str, ir: bool) -> Self {
```
Comment (duesee) [inline tasks/src/tasks/authenticate.rs:80]: I think we can call this just "plain".

**Review by duesee — COMMENTED** [review]
Hey Clement, some preliminary comments :-)

File: tasks/src/lib.rs [hunk 5]
```diff
@@ -136,6 +139,40 @@ impl Scheduler {
...
+    /// Run the [`LogoutTask`].
+    pub fn logout(&mut self) -> TaskRunner<LogoutTask> {
+        self.run_task(LogoutTask::new())
+    }
+
+    /// Run the [`NoOpTask`].
+    pub fn noop(&mut self) -> TaskRunner<NoOpTask> {
+        self.run_task(NoOpTask::new())
+    }
+
```
[the remark this answers is not in the document]
Comment (soywod) [reply tasks/src/lib.rs]: It makes sense, then I will make the high-level API use directly tasks instead.

File: tasks/Cargo.toml [hunk 6]
```diff
@@ -7,6 +7,10 @@ license = "MIT OR Apache-2.0"
 [dependencies]
 imap-flow = { path = ".." }
 imap-types = "2.0.0"
+static_assertions = "1.1.0"
 tag-generator = { path = "../tag-generator" }
 thiserror = "1.0.58"
+tokio = { version = "1.37.0", features = ["net"] }
```
Comment (jakoschiko) [inline tasks/Cargo.toml:13]: Do we need tokio as non-dev dependency? I thought `Scheduler` is sans I/O now.

File: tasks/examples/client-tasks.rs [hunk 7]
```diff
@@ -16,25 +16,12 @@ async fn main() {
...
-    }
+    let capabilities = stream
+        .progress(scheduler.run_task(CapabilityTask::new()))
+        .await
+        .unwrap()
+        .unwrap();
+    println!("capabilities: {capabilities:#?}");
 
     let handle2 = scheduler.enqueue_task(AuthenticateTask::plain("alice", "pa²²w0rd", true));
     let handle3 = scheduler.enqueue_task(LogoutTask::default());
```
Comment (jakoschiko) [inline tasks/examples/client-tasks.rs:27]: `handle1` is missing now. Let's rename the variables to `authenticate_handle` and `logout_handle`.

File: tasks/src/lib.rs [hunk 8]
```diff
@@ -342,6 +355,42 @@ pub enum SchedulerError {
...
+        self.scheduler.enqueue_input(bytes);
+    }
+
+    fn progress(&mut self) -> Result<Self::Event, FlowInterrupt<Self::Error>> {
+        loop {
+            match self.scheduler.progress()? {
+                SchedulerEvent::TaskFinished(mut token) => {
+                    if let Some(output) = self.handle.resolve(&mut token) {
+                        break Ok(output);
+                    }
```
Comment (jakoschiko) [inline tasks/src/lib.rs:381]: In the long run we should find a better idea how to deal with unexpected tokens here. But for now, what do you think about addding a log statement with WARN level? `imap-flow` already has `tracing` as logging dependency.

File: tasks/src/lib.rs [hunk 9]
```diff
@@ -342,6 +355,42 @@ pub enum SchedulerError {
...
+                }
+                SchedulerEvent::Unsolicited(unsolicited) => {
+                    if let Response::Status(Status::Bye(bye)) = unsolicited {
+                        let err = SchedulerError::UnexpectedByeResponse(bye);
+                        break Err(FlowInterrupt::Error(err));
+                    } else {
+                        println!("unsolicited: {unsolicited:?}");
```
Comment (jakoschiko) [inline tasks/src/lib.rs:388]: Let's better change this `println` to a log statement with WARN level.

**Comment by jakoschiko** [comment]
Hello! Thanks for progressing this PR. We were just discussing it and stumbled upon a design issue that is created by having `enqueue_task` and `run_task` on the `Scheduler`. Let's look at this example:

```rust
let handle = scheduler.enqueue_task(AuthenticateTask::plain("alice", "pa²²w0rd", true));

let capabilities = stream
    .progress(scheduler.run_task(CapabilityTask::new()))
    .await // This will silently ignore the result of the enqueued AuthenticateTask
    .unwrap()
    .unwrap();
```

Having both methods on the same type might lead to accidental misusage. Maybe you were right about introducing a dedicated type called `Resolver` (or any other name).

- `Scheduler` is more low-level and provides only `enqueue_task`. Its usage is complicated but rather flexible.
- `Resolver` is more high-level and provides only `run_task`. Its usage is easy but rather restrictive.

However, these are long-term thoughts. If you are happy with the current state then let's keep it for now and improve the design later. I'm sure I'll change my mind about this a couple of times :p

**Review by duesee — APPROVED** [review]
Awesome, thanks! As soon as the remaining discussions are fixed we should quickly merge :-) I'm already accepting, thanks for the perseverence!

File: tasks/Cargo.toml (hunk 6 above)
Comment (soywod) [reply tasks/Cargo.toml:13]: No we don't, it's just dead code from refactors. I will remove it, nice catch!

**Comment by soywod** [comment]
> Having both methods on the same type might lead to accidental misusage. Maybe you were right about introducing a dedicated type called `Resolver` (or any other name).

Indeed, I makes sense to separate them to avoid misuage. I initiated a new module `resolver`, let me know what you think about it.

Ready for another review!

File: tasks/src/resolver.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    fn progress(&mut self) -> Result<Self::Event, FlowInterrupt<Self::Error>> {
+        self.scheduler.progress()
+    }
+}
+
+impl Resolver {
+    /// Create a new resolver.
+    pub fn new(flow: ClientFlow) -> Self {
```
Comment (jakoschiko) [inline tasks/src/resolver.rs:27]: I like that the `Scheduler` is just an implementation of `Resolver` and does not appear in its API.

File: tasks/src/resolver.rs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        Self {
+            scheduler: Scheduler::new(flow),
+        }
+    }
+
+    /// Enqueue a [`Task`] for immediate resolution.
+    pub fn run_task<T: Task>(&mut self, task: T) -> TaskResolver<T> {
```
Comment (jakoschiko) [inline tasks/src/resolver.rs:34]: I'm not quite happy with the current names `Resolver`, `Resolver::run_task` and `TaskResolver`.

Some alternatives:

- `Resolver`, `Resolver::resolve`, `ResolvingTask`
- `Runner`, `Runner::run`, `RunningTask`

But I'm just nitpicking. I you like we can keep the current names for now.

File: tasks/src/resolver.rs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    fn progress(&mut self) -> Result<Self::Event, FlowInterrupt<Self::Error>> {
+        loop {
+            match self.resolver.progress()? {
+                SchedulerEvent::TaskFinished(mut token) => {
+                    if let Some(output) = self.handle.resolve(&mut token) {
+                        break Ok(output);
+                    } else {
+                        warn!("received unexpected task token: {token:?}")
```
Comment (jakoschiko) [inline tasks/src/resolver.rs:64]: When using `tracing` it's more common to separate message and data. 

```suggestion
                        warn!(?token, "received unexpected task token")
```

File: tasks/src/resolver.rs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                    }
+                }
+                SchedulerEvent::Unsolicited(unsolicited) => {
+                    if let Response::Status(Status::Bye(bye)) = unsolicited {
+                        let err = SchedulerError::UnexpectedByeResponse(bye);
+                        break Err(FlowInterrupt::Error(err));
+                    } else {
+                        warn!("received unsolicited: {unsolicited:?}");
```
Comment (jakoschiko) [inline tasks/src/resolver.rs:72]:
```suggestion
                        warn!(?unsolicited, "received unsolicited");
```

**Review by jakoschiko — APPROVED** [review]

**Comment by jakoschiko** [comment]
Thanks again! I added some more comments, but only minor issues, you can choose whether you want to fix them.

If you are ready, we can merge it.

File: tasks/src/resolver.rs (hunk 11 above)
Comment (soywod) [reply tasks/src/resolver.rs:34]: I prefer resolve over run, because it gives this information of running a task till the end. So I opted for the first one :ok_hand:

**Comment by soywod** [comment]
> Thanks again! I added some more comments, but only minor issues, you can choose whether you want to fix them.

Fixes pushed, waiting for the CI. If no more comment then it can be merged straight! I will open another PR with all other tasks.

Thank you :slightly_smiling_face:

File: tasks/src/resolver.rs (hunk 10 above)
Comment (soywod) [reply tasks/src/resolver.rs:27]: That said, the resolver still uses scheduler events and errors, is it OK? Should type aliases or struct/enum wrappers be defined?

Comment (jakoschiko) [reply tasks/src/resolver.rs:27]: It's okay. There are different alternatives how we could proceed and I don't want to decide it now. The solution should probably also include the unexpected cases that we log at the moment.

**Review by jakoschiko — APPROVED** [review]
